File: app/src/main/java/fr/skyost/timetable/widget/TodayWidgetDateManager.java
package fr.skyost.timetable.widget;

import org.joda.time.LocalDate;

/**
 * The today's widget date manager.
 */

public class TodayWidgetDateManager {

	/**
	 * Relative day (from today).
	 */

	private int relativeDay = 0;

	/**
	 * The instance holder.
	 */

	private static class Holder {

		/**
		 * The instance.
		 */

		private static final TodayWidgetDateManager INSTANCE = new TodayWidgetDateManager();

	}

	/**
	 * Returns the class instance.
	 *
	 * @return The class instance.
	 */

	public static TodayWidgetDateManager getInstance() {
		return Holder.INSTANCE;
	}

	/**
	 * Returns the relative day.
	 *
	 * @return The relative day.
	 */

	public int getRelativeDay() {
		return relativeDay;
	}

	/**
	 * Adds a day to the relative day count.
	 */

	public void plusRelativeDay() {
		this.relativeDay++;
	}

	/**
	 * Subtracts a day to the relative day count.
	 */

	public void minusRelativeDay() {
		this.relativeDay--;
	}

	/**
	 * Sets the relative day.
	 *
	 * @param relativeDay The relative day.
	 */

	public void setRelativeDay(final int relativeDay) {
		this.relativeDay = relativeDay;
	}

	/**
	 * Returns the current absolute day.
	 *
	 * @return The current absolute day.
	 */

	public LocalDate getAbsoluteDay() {
		return LocalDate.now().plusDays(relativeDay);
	}

}

File: android/app/src/main/kotlin/fr/skyost/timetable/widget/TodayWidgetDateManager.kt
package fr.skyost.timetable.widget

import java.time.LocalDate

/**
 * Statically holds a relative day per widget.
 */
class TodayWidgetDateManager private constructor() {
    companion object {
        /**
         * The holders map.
         */
        private val holders: MutableMap<Int, Int> = HashMap()

        /**
         * Returns the class instance.
         *
         * @param widgetId The widget id.
         *
         * @return The class instance.
         */
        fun getRelativeDay(widgetId: Int): Int {
            var relativeDay = holders[widgetId]
            if (relativeDay == null) {
                relativeDay = 0
                changeRelativeDay(widgetId, relativeDay)
            }
            return relativeDay
        }

        /**
         * Returns the current absolute day.
         *
         * @param widgetId The widget id.
         *
         * @return The current absolute day.
         */
        fun resolveAbsoluteDay(widgetId: Int): LocalDate {
            return LocalDate.now().plusDays(getRelativeDay(widgetId).toLong())
        }

        /**
         * Returns the class instance.
         *
         * @param widgetId The widget id.
         * @param relativeDay The new relative day?
         *
         * @return The class instance.
         */
        fun changeRelativeDay(widgetId: Int, relativeDay: Int) {
            holders[widgetId] = relativeDay
        }
    }
}
